### PR TITLE
HDF5 Driver Update

### PR DESCRIPTION
**Pull request by mrossinek**
### Summary

Refactors the `HDF5Driver` to be aligned with #461.
It will also support loading legacy QMolecule files stored in HDF5 files.

### CODE REVIEW

File: qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py
Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py:83]: Yes I have this planned. However, doing this is currently still blocked by #552 and #521. I figured I'll add another PR after those or add it into this PR if we merge the others soon enough :+1:

File: qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py
Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py:79]: Right I forgot to add this utility method. Will do that :+1: 
Release note will be added, too

File: qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py
Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py:103]: Yes, with #554 this is no longer needed.

I will update the check against the property type, too

**Comment by mrossinek** [comment]
I rebased this PR on #554, blocking it until that is merged.

**Review by dlasecki — APPROVED** [review]

File: qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py
Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py:124]: Tracked in separate issue: https://github.com/Qiskit/qiskit-nature/issues/561

File: qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py
Comment (woodsp-ibm) [inline qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py:79]: I am wondering about this being an instance method and the default overwriting of the same file seems destructive by default and of concern if anything goes wrong. Would this be better as a static method or function in this same module that takes the input file name, an optional output file name and replace False by default. So it becomes more of a conversion function that can be used that is not directly a driver. By default the output file name could be the input file name part with _new appended. E.g I give it "h2_631g.hdf5" which is a legacy hdf5 and it would create, by default, "h2_631g_new.hdf5", unless I give it an explicit name. If the output already exists, for whatever, it would require them to set replace to True. Buyer beware I guess if you give the output file the same name as the input and replace True. But at least it would not be the default behavior where they may lose their input file with it being overwritten and intended to keep it. If the input is not a legacy hdf5 file the raise an error indicating  that etc. If you prefer to keep this as an instance method, supplying the source file to be converted via the constructor I the behavior could still be the same.

Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py:79]: I have updated the default value of `replace` and the HDF5 suffix handling.

However, I kept `convert()` as an instance method. My reasoning is that a user may encounter the warning of using a legacy file at runtime where they already instantiated the driver object. They can then simply run `driver.convert()` to update the legacy file.

File: qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py
Comment (woodsp-ibm) [inline qiskit_nature/drivers/second_quantization/hdf5d/hdf5driver.py:84]: I think text needs to be updated with the updated default naming

**Review by woodsp-ibm — APPROVED** [review]

**Review by pbark — APPROVED** [review]
LGTM! Thanks @mrossinek!

**Pull request merged**
